processor/src/main/java/com/linecorp/decaton/processor/runtime/ProcessorSubscription.java

```diff
@@ -179,6 +179,11 @@ public void run() {
             consumer.subscribe(subscribeTopics(), new ConsumerRebalanceListener() {
                 @Override
                 public void onPartitionsRevoked(Collection<TopicPartition> partitions) {
+                    // KafkaConsumer#close has been changed to invoke onPartitionRevoked since Kafka 2.4.0.
+                    // Since we're doing cleanup procedure on shutdown individually so just immediately return if terminated
+                    if (terminated.get()) {
```

Comment:
Hmm sounds true.... thanks for pointing out.

Instead of adding another flag like `isClosing`, how about storing subscription's state in `updateState` and checking it? (`if (state == State.SHUTTING_DOWN) {...`)

Reply:
Sounds workable.
We would have few concerns thought;
- As of current version kafka states that the rebalance listener is invoked only during active call of `Consumer#poll`(see javadoc of `KafkaConsumer#subscribe`, which is already a tiny lie since it also gets invoked during `Consumer#close`....) so it should be safe to use `SHUTTING_DOWN` state flag as-is. However if in future, either we add a new operation in between `updateState(SHUTTING_DOWN)` and `consumer.close()`, we would have to introduce another State variant which is pointless for users but is used only for internal flow control.


However, If we can get rid of `terminated` flag with it by introducing another state that express a state between `while (...)` and the head of finally block (such as `REQUESTING_TERMINATE`?) that would be the most cleaner (but would be a trade-off with simper states expression (for users use) vs more variants for internal state management).

Reply:
Hm, I see.
I think introducing another state to get rid of `terminated` has one more concern that it will violates this contract: https://github.com/line/decaton/blob/master/processor/src/main/java/com/linecorp/decaton/processor/SubscriptionStateListener.java#L83

Then introducing another flag as first suggestion sounds most appropriate for me.

Reply:
I think add an integration test to check expected state transition can cover this?
> preferred to have a test case proving that it won't happen

- partial invocation of rebalance listener never happen => i.e. `onChange(REBALANCING)` and `onChange(RUNNING)` always be invoked in this order in pair

Reply:
> I think add an integration test to check expected state transition can cover this?

That's one good-to-have kind of test indeed.

Reply:
>  I think introducing another state to get rid of terminated has one more concern that it will violates this contract: https://github.com/line/decaton/blob/master/processor/src/main/java/com/linecorp/decaton/processor/SubscriptionStateListener.java#L83

good point, I haven't came up with it.

Reply:
Added a test about state transition: https://github.com/line/decaton/pull/48/files#diff-ffc890431d9725fc4d648d1543fce5d0R38